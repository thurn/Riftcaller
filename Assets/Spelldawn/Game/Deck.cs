// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Collections.Generic;
using System.Linq;
using Spelldawn.Protos;
using Spelldawn.Services;
using Spelldawn.Utils;
using UnityEngine;

#nullable enable

namespace Spelldawn.Game
{
  public sealed class Deck : StackObjectDisplay
  {
    [SerializeField] Registry _registry = null!;
    [SerializeField] PlayerName _owner;
    bool _canTakeDrawCardAction;

    bool Clickable => _owner == PlayerName.User &&
                      _canTakeDrawCardAction &&
                      _registry.CapabilityService.CanExecuteAction(ClientAction.ActionOneofCase.DrawCard);    
    
    protected override Registry Registry => _registry;

    protected override GameContext DefaultGameContext() => GameContext.Deck;

    public void RenderDeckView(DeckView deckView)
    {
      _canTakeDrawCardAction = deckView.CanTakeDrawCardAction;
      
      foreach (var obj in AllObjects)
      {
        var spriteRenderer = obj.GetComponent<SpriteRenderer>();
        if (spriteRenderer)
        {
          // Update card back images for placeholder cards
          _registry.AssetService.AssignSprite(spriteRenderer, deckView.CardBack);
        }
      }
    }

    public IEnumerable<Card> Cards() => AllObjects.OfType<Card>();

    protected override void LongPress()
    {
      StartCoroutine(_registry.LongPressCardBrowser.BrowseCards(this));
    }
    
    public override bool CanHandleMouseEvents() => true;
    
    public override void MouseUp()
    {
      if (Clickable)
      {
        if (_registry.CapabilityService.CanExecuteAction(ClientAction.ActionOneofCase.DrawCard))
        {
          _registry.ActionService.HandleAction(new ClientAction
          {
            DrawCard = new DrawCardAction()
          });
        }
        else
        {
          LogUtils.Log("Ignoring click on deck, cannot currently draw a card");
        }
      }
    }
  }
}