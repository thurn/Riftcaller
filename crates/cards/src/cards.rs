// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Contains the definitions for all cards in the game.

use assets::rexard_images;
use assets::rexard_images::RexardPack;
use card_helpers::identity_cost;
use data::card_definition::{CardConfig, CardDefinition};
use data::card_name::CardName;
use data::primitives::{CardType, Rarity, School, Side};

pub mod artifacts;
pub mod canonical_game;
pub mod champion_identities;
pub mod champion_spells;
pub mod minions;
pub mod overlord_identities;
pub mod overlord_spells;
pub mod projects;
pub mod schemes;
pub mod weapons;

pub fn no_identity(name: CardName, side: Side, school: School) -> CardDefinition {
    CardDefinition {
        name,
        sets: vec![],
        cost: identity_cost(),
        image: rexard_images::get(RexardPack::MonstersAvatars, "22"),
        card_type: CardType::Identity,
        side,
        school,
        rarity: Rarity::Common,
        abilities: vec![],
        config: CardConfig::default(),
    }
}
