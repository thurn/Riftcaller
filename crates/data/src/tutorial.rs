// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::collections::HashSet;

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq, Hash)]
pub enum TutorialMessageKey {
    DeckEditor,
}

/// Data model for the player's progress through the game's tutorial
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct TutorialData {
    /// Which tutorial messages has the user seen?
    pub seen: HashSet<TutorialMessageKey>,
}

impl TutorialData {
    pub fn new() -> Self {
        Self::default()
    }
}
