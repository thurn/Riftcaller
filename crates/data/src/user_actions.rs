// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use serde::{Deserialize, Serialize};

use crate::card_name::CardName;
use crate::game_actions::GameAction;
use crate::player_name::{NamedPlayer, PlayerId};
use crate::primitives::{ActionCount, DeckIndex, GameId, ManaValue, PointsValue, School, Side};

#[derive(Eq, PartialEq, Hash, Debug, Copy, Clone, Default, Serialize, Deserialize)]
pub struct NewGameDebugOptions {
    /// If true, all game events will be non-random
    pub deterministic: bool,
    /// Explicitly set the ID for this game
    pub override_game_id: Option<GameId>,
}

/// Action to initiate a new game
#[derive(Eq, PartialEq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub struct NewGameAction {
    /// Deck to use for this game
    pub deck_index: DeckIndex,
    /// Opponent to play against
    pub opponent: PlayerId,
    /// Debug configuration for this game
    pub debug_options: Option<NewGameDebugOptions>,
}

/// Actions that can be taken from the debug panel, should not be exposed in
/// production.
#[derive(Eq, PartialEq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum DebugAction {
    /// Creates a new game with ID 0, using the canonical decklist for [Side],
    /// playing against an opponent who will take no actions. Overwrites the
    /// current player's player data with the canonical decklists.
    NewGame(Side),

    /// Adds the current player to the game with ID 0, overwriting the non-human
    /// player in this game. Overwrites the current player's player data
    /// with the canonical decklists.
    JoinGame,

    /// Swaps which side the current player is playing as in their current game.
    FlipViewpoint,

    AddMana(ManaValue),
    AddActionPoints(ActionCount),
    AddScore(PointsValue),
    SaveState(u64),
    LoadState(u64),
    SetNamedPlayer(Side, NamedPlayer),

    /// Gives the player copies of every card
    FullCollection,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, Eq, PartialEq, Hash)]
pub enum DeckEditorAction {
    /// Create a new deck for the current player
    CreateDeck(Side, School),
    /// Add one copy of a card to a deck
    AddToDeck(CardName, DeckIndex),
    /// Remove one copy of a card from a deck
    RemoveFromDeck(CardName, DeckIndex),
}

impl From<DeckEditorAction> for UserAction {
    fn from(a: DeckEditorAction) -> Self {
        UserAction::DeckEditorAction(a)
    }
}

/// All possible action payloads that can be sent from a client
#[derive(Debug, Clone, Copy, Serialize, Deserialize, Eq, PartialEq, Hash)]
pub enum UserAction {
    NewGame(NewGameAction),
    Debug(DebugAction),
    GameAction(GameAction),
    DeckEditorAction(DeckEditorAction),
}
