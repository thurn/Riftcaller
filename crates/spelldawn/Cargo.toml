[package]
name = "spelldawn"
version = "0.0.0"
homepage = "https://github.com/thurn/spelldawn"
description = "Spelldawn: An asymmetric trading card game"
license = "Apache-2.0"
default-run = "spelldawn"
edition = "2021"

[[bin]]
name = "spelldawn"
path = "src/main.rs"
test = false
doctest = false

[dependencies]
cards_all = { path = "../cards/cards_all", version = "0.0.0" }
protos = { path = "../protos", version = "0.0.0" }
server = { path = "../server", version = "0.0.0" }

anyhow = "1.0.58"
tokio = { version = "1.19.2", features = ["macros", "rt-multi-thread", "sync", "time"] }
tonic = { version = "0.7.2", features = ["compression"] }
tonic-web = "0.3.0"
tracing = "0.1.35"
tracing-subscriber = "0.3.14"

[dev-dependencies]
rusty-hook = "0.11.2"
