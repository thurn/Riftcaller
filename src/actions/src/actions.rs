// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Contains functions for responding to user-initiated game actions received
//! from the client. The `handle_user_action` function is the primary
//! entry-point into the rules engine.

pub mod legal_actions;

use anyhow::Result;
use game_data::card_definition::AbilityType;
use game_data::card_state::CardPosition;
use game_data::delegates::{
    AbilityActivated, ActivateAbilityEvent, CardPlayed, CastCardEvent, DrawCardActionEvent,
};
use game_data::game::{GamePhase, GameState, HistoryEntry, HistoryEvent, MulliganDecision};
use game_data::game_actions::{CardTarget, GameAction, PromptAction};
use game_data::primitives::{AbilityId, CardId, RoomId, Side};
use game_data::updates::{GameUpdate, InitiatedBy};
use raids::RaidDataExt;
use rules::mana::ManaPurpose;
use rules::{card_prompt, dispatch, flags, mana, mutations, queries};
use tracing::{debug, instrument};
use with_error::{fail, verify, WithError};

/// Top level dispatch function responsible for mutating [GameState] in response
/// to all [GameAction]s
pub fn handle_game_action(
    game: &mut GameState,
    user_side: Side,
    action: &GameAction,
) -> Result<()> {
    match action {
        GameAction::PromptAction(prompt_action) => {
            handle_prompt_action(game, user_side, *prompt_action)
        }
        GameAction::Resign => handle_resign_action(game, user_side),
        GameAction::GainMana => gain_mana_action(game, user_side),
        GameAction::DrawCard => draw_card_action(game, user_side),
        GameAction::PlayCard(card_id, target) => {
            play_card_action(game, user_side, *card_id, *target)
        }
        GameAction::ActivateAbility(ability_id, target) => {
            activate_ability_action(game, user_side, *ability_id, *target)
        }
        GameAction::UnveilCard(card_id) => unveil_action(game, user_side, *card_id),
        GameAction::InitiateRaid(room_id) => {
            raids::handle_initiate_action(game, user_side, *room_id)
        }
        GameAction::LevelUpRoom(room_id) => level_up_room_action(game, user_side, *room_id),
        GameAction::SpendActionPoint => spend_action_point_action(game, user_side),
    }?;

    Ok(())
}

/// Returns true if the indicated player currently has a legal game action
/// available to them.
pub fn has_priority(game: &GameState, side: Side) -> bool {
    match &game.info.phase {
        GamePhase::ResolveMulligans(_) => return flags::can_make_mulligan_decision(game, side),
        GamePhase::GameOver { .. } => return false,
        _ => {}
    };

    match &game.info.raid {
        Some(raid) => side == raid.phase().active_side(),
        None => side == game.info.turn.side,
    }
}

fn handle_resign_action(game: &mut GameState, side: Side) -> Result<()> {
    debug!(?side, "Applying resign action");
    if !matches!(game.info.phase, GamePhase::GameOver { .. }) {
        mutations::game_over(game, side.opponent())?;
    }
    Ok(())
}

/// Handles a choice to keep or mulligan an opening hand
fn handle_mulligan_decision(
    game: &mut GameState,
    user_side: Side,
    decision: MulliganDecision,
) -> Result<()> {
    verify!(
        flags::can_make_mulligan_decision(game, user_side),
        "Cannot make mulligan decision for {:?}",
        user_side
    );

    debug!(?user_side, ?decision, "Applying mulligan action");
    let mut mulligans = match &mut game.info.phase {
        GamePhase::ResolveMulligans(mulligans) => mulligans,
        _ => fail!("Incorrect game phase"),
    };

    match user_side {
        Side::Overlord => mulligans.overlord = Some(decision),
        Side::Champion => mulligans.champion = Some(decision),
    }

    let hand = game.card_list_for_position(user_side, CardPosition::Hand(user_side));
    match decision {
        MulliganDecision::Keep => {}
        MulliganDecision::Mulligan => {
            mutations::shuffle_into_deck(game, user_side, &hand)?;
            mutations::draw_cards(game, user_side, 5)?;
        }
    }

    mutations::check_start_game(game)?;

    Ok(())
}

/// The basic game action to draw a card during your turn by spending one
/// action.
#[instrument(skip(game))]
fn draw_card_action(game: &mut GameState, user_side: Side) -> Result<()> {
    verify!(
        flags::can_take_draw_card_action(game, user_side),
        "Cannot draw card for {:?}",
        user_side
    );

    debug!(?user_side, "Applying draw card action");
    mutations::spend_action_points(game, user_side, 1)?;
    let cards = mutations::draw_cards(game, user_side, 1)?;
    if let Some(card_id) = cards.get(0) {
        dispatch::invoke_event(game, DrawCardActionEvent(*card_id))?;
    }

    mutations::check_end_turn(game)?;
    Ok(())
}

/// The basic game action to play a card during your turn. Spends the resource
/// cost for a card, resolves its effects, and then moves it to the appropriate
/// new [CardPosition]. Spell, Weapon, and Artifact cards are immediately
/// revealed when played.
#[instrument(skip(game))]
fn play_card_action(
    game: &mut GameState,
    user_side: Side,
    card_id: CardId,
    target: CardTarget,
) -> Result<()> {
    verify!(
        flags::can_take_play_card_action(game, user_side, card_id, target),
        "Cannot play card {:?}",
        card_id
    );

    let card = game.card(card_id);
    debug!(?card.name, ?user_side, ?card_id, ?target, "Applying play card action");

    let definition = rules::get(card.name);
    mutations::move_card(game, card_id, CardPosition::Played(user_side, target))?;

    let actions = queries::action_cost(game, card_id);
    mutations::spend_action_points(game, user_side, actions)?;

    if flags::enters_play_face_up(game, card_id) {
        let amount = queries::mana_cost(game, card_id).with_error(|| "Card has no mana cost")?;
        mana::spend(game, user_side, ManaPurpose::PayForCard(card_id), amount)?;
        if let Some(custom_cost) = &definition.cost.custom_cost {
            (custom_cost.pay)(game, card_id)?;
        }
        game.card_mut(card_id).turn_face_up();
        game.record_update(|| GameUpdate::PlayCardFaceUp(user_side, card_id));
    }

    dispatch::invoke_event(game, CastCardEvent(CardPlayed { card_id, target }))?;
    mutations::move_card(
        game,
        card_id,
        queries::played_position(game, user_side, card_id, target)?,
    )?;

    game.history
        .push(HistoryEntry { turn: game.info.turn, event: HistoryEvent::PlayedCard(card_id) });
    mutations::check_end_turn(game)?;
    Ok(())
}

/// The basic game action to activate an ability of a card in play.
#[instrument(skip(game))]
fn activate_ability_action(
    game: &mut GameState,
    user_side: Side,
    ability_id: AbilityId,
    target: CardTarget,
) -> Result<()> {
    verify!(
        flags::can_take_activate_ability_action(game, user_side, ability_id, target),
        "Cannot activate ability {:?}",
        ability_id
    );

    game.ability_state.entry(ability_id).or_default().currently_resolving = true;
    let card = game.card(ability_id.card_id);

    debug!(?card.name, ?user_side, ?ability_id, "Applying activate ability action");

    let cost = match &rules::get(card.name).ability(ability_id.index).ability_type {
        AbilityType::Activated(cost, _) => cost,
        _ => fail!("Ability is not an activated ability"),
    };

    mutations::spend_action_points(game, user_side, cost.actions)?;
    if let Some(mana) = queries::ability_mana_cost(game, ability_id) {
        mana::spend(game, user_side, ManaPurpose::ActivateAbility(ability_id), mana)?;
    }

    if let Some(custom_cost) = &cost.custom_cost {
        (custom_cost.pay)(game, ability_id)?;
    }

    game.record_update(|| GameUpdate::AbilityActivated(user_side, ability_id));
    dispatch::invoke_event(game, ActivateAbilityEvent(AbilityActivated { ability_id, target }))?;

    game.ability_state.entry(ability_id).or_default().currently_resolving = false;
    mutations::check_end_turn(game)?;
    Ok(())
}

/// The basic game action to unveil a project card in play.
#[instrument(skip(game))]
fn unveil_action(game: &mut GameState, user_side: Side, card_id: CardId) -> Result<()> {
    verify!(
        flags::can_take_unveil_card_action(game, user_side, card_id),
        "Cannot unveil card {:?}",
        card_id
    );

    mutations::unveil_card(game, card_id)?;
    Ok(())
}

/// The basic game action to gain 1 mana during your turn by spending one
/// action.
#[instrument(skip(game))]
fn gain_mana_action(game: &mut GameState, user_side: Side) -> Result<()> {
    verify!(
        flags::can_take_gain_mana_action(game, user_side),
        "Cannot gain mana for {:?}",
        user_side
    );

    debug!(?user_side, "Applying gain mana action");
    mutations::spend_action_points(game, user_side, 1)?;
    mana::gain(game, user_side, 1);
    mutations::check_end_turn(game)?;
    Ok(())
}

fn level_up_room_action(game: &mut GameState, user_side: Side, room_id: RoomId) -> Result<()> {
    verify!(
        flags::can_take_level_up_room_action(game, user_side, room_id),
        "Cannot level up room for {:?}",
        user_side
    );
    debug!(?user_side, "Applying level up room action");
    mutations::spend_action_points(game, user_side, 1)?;
    mana::spend(game, user_side, ManaPurpose::LevelUpRoom(room_id), 1)?;
    game.record_update(|| GameUpdate::LevelUpRoom(room_id, InitiatedBy::GameAction));
    mutations::level_up_room(game, room_id)?;
    mutations::check_end_turn(game)?;
    Ok(())
}

fn spend_action_point_action(game: &mut GameState, user_side: Side) -> Result<()> {
    verify!(flags::in_main_phase(game, user_side), "Cannot spend action point for {:?}", user_side);
    mutations::spend_action_points(game, user_side, 1)?;
    mutations::check_end_turn(game)?;
    Ok(())
}

/// Handles a [PromptAction] for the `user_side` player and then removes it from
/// the queue.
fn handle_prompt_action(game: &mut GameState, user_side: Side, action: PromptAction) -> Result<()> {
    if let Some(prompt) = &game.player(user_side).card_prompt_queue.get(0) {
        verify!(
            prompt.responses.iter().any(|p| *p == action),
            "Unexpected action {:?} received",
            action
        );

        game.player_mut(user_side).card_prompt_queue.remove(0);
    } else if matches!(action, PromptAction::CardAction(_)) {
        fail!("Received action with no matching prompt {:?}", action);
    }

    match action {
        PromptAction::MulliganDecision(mulligan) => {
            handle_mulligan_decision(game, user_side, mulligan)
        }
        PromptAction::SummonAction(_)
        | PromptAction::EncounterAction(_)
        | PromptAction::AccessPhaseAction(_) => raids::handle_action(game, user_side, action),
        PromptAction::CardAction(card_action) => card_prompt::handle(game, user_side, card_action),
    }
}
