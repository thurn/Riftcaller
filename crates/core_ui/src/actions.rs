// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::collections::HashMap;
use std::fmt::Debug;

use data::adventure_action::AdventureAction;
use data::game_actions::{GameAction, PromptAction};
use data::user_actions::{DebugAction, UserAction};
use protos::spelldawn::client_action::Action;
use protos::spelldawn::game_command::Command;
use protos::spelldawn::{ClientAction, CommandList, GameCommand, StandardAction};
use serde_json::ser;

/// Represents an action that can be performed in the user interface. Initiating
/// a server request and performing an immediate client update are both
/// supported forms of action.
pub trait InterfaceAction {
    fn as_client_action(&self) -> Action;

    /// Converts an [InterfaceAction] into a [ClientAction].
    fn build(&self) -> ClientAction {
        ClientAction { action: Some(self.as_client_action()) }
    }
}

impl InterfaceAction for Action {
    fn as_client_action(&self) -> Action {
        self.clone()
    }
}

impl InterfaceAction for StandardAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(self.clone())
    }
}

/// Marker struct for when no action is desired.
#[derive(Debug)]
pub struct NoAction {}

impl InterfaceAction for NoAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction::default())
    }
}

impl InterfaceAction for UserAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: payload(*self),
            update: None,
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for DebugAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: payload(UserAction::Debug(*self)),
            update: None,
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for GameAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: payload(UserAction::GameAction(*self)),
            update: None,
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for PromptAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: payload(UserAction::GameAction(GameAction::PromptAction(*self))),
            update: None,
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for AdventureAction {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: payload(UserAction::AdventureAction(*self)),
            update: None,
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for Command {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: vec![],
            update: Some(command_list(vec![self.clone()])),
            request_fields: HashMap::new(),
        })
    }
}

impl InterfaceAction for Vec<Command> {
    fn as_client_action(&self) -> Action {
        Action::StandardAction(StandardAction {
            payload: vec![],
            update: Some(command_list(self.clone())),
            request_fields: HashMap::new(),
        })
    }
}

/// Returns a StandardAction to apply an optimistic UI update via a series on
/// commands on click, followed by a regular game action specified as an
/// InterfaceAction.
///
/// The provided InterfaceAction must map to a StandardAction, or else this
/// function will panic.
pub fn with_optimistic_update(
    update: Vec<Command>,
    action: impl InterfaceAction + 'static,
) -> Action {
    let payload = match action.as_client_action() {
        Action::StandardAction(standard_action) => standard_action.payload,
        _ => panic!("Expected StandardAction"),
    };

    Action::StandardAction(StandardAction {
        payload,
        update: Some(command_list(update)),
        request_fields: HashMap::new(),
    })
}

pub fn payload(action: UserAction) -> Vec<u8> {
    ser::to_vec(&action).expect("Serialization failed")
}

pub fn command_list(commands: Vec<Command>) -> CommandList {
    CommandList {
        commands: commands.into_iter().map(|c| GameCommand { command: Some(c) }).collect(),
    }
}
