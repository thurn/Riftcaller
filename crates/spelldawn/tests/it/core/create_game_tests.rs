// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use cards::initialize;
use core_ui::actions::InterfaceAction;
use data::card_name::CardName;
use data::deck::Deck;
use data::game::MulliganDecision;
use data::game_actions::{GameAction, PromptAction};
use data::player_data::PlayerData;
use data::player_name::PlayerId;
use data::primitives::{DeckIndex, GameId, Side};
use data::tutorial::TutorialData;
use data::user_actions::{NewGameAction, NewGameDebugOptions, UserAction};
use insta::assert_snapshot;
use maplit::hashmap;
use protos::spelldawn::PlayerName;
use test_utils::client::TestSession;
use test_utils::client_interface::HasText;
use test_utils::fake_database::FakeDatabase;
use test_utils::summarize::Summary;
use test_utils::*;

static OVERLORD_DECK: DeckIndex = DeckIndex { value: 0 };
static CHAMPION_DECK: DeckIndex = DeckIndex { value: 1 };

#[test]
fn create_new_game() {
    let (game_id, overlord_id, champion_id) = generate_ids();
    let mut session = make_overlord_test_session(game_id, overlord_id, champion_id);
    let response = session.perform_action(
        UserAction::NewGame(NewGameAction {
            deck_index: OVERLORD_DECK,
            opponent: session.opponent_id(),
            debug_options: Some(NewGameDebugOptions {
                deterministic: true,
                ..NewGameDebugOptions::default()
            }),
        })
        .as_client_action(),
        session.user_id(),
    );
    assert_snapshot!(Summary::run(&response));
}

#[test]
fn connect_to_new_game() {
    let (game_id, overlord_id, champion_id) = generate_ids();
    let mut session = make_overlord_test_session(game_id, overlord_id, champion_id);
    initiate_game(&mut session);

    let response = session.connect(overlord_id);
    assert!(session.user.interface.controls().has_text("Keep"));
    assert!(session.user.interface.controls().has_text("Mulligan"));
    assert_eq!(5, session.user.cards.revealed_cards().len());
    assert_eq!(5, session.user.cards.hand(PlayerName::Opponent).len());

    assert_snapshot!(Summary::run(&response));
}

#[test]
fn mulligan_legal_actions() {
    let (game_id, overlord_id, champion_id) = generate_ids();
    let mut session = make_overlord_test_session(game_id, overlord_id, champion_id);
    initiate_game(&mut session);

    assert_contents_equal(
        session.legal_actions(Side::Overlord),
        vec![
            GameAction::PromptAction(PromptAction::MulliganDecision(MulliganDecision::Mulligan)),
            GameAction::PromptAction(PromptAction::MulliganDecision(MulliganDecision::Keep)),
        ],
    );
    assert_contents_equal(
        session.legal_actions(Side::Champion),
        vec![
            GameAction::PromptAction(PromptAction::MulliganDecision(MulliganDecision::Mulligan)),
            GameAction::PromptAction(PromptAction::MulliganDecision(MulliganDecision::Keep)),
        ],
    );

    session.click_on(overlord_id, "Keep");
    assert!(session.legal_actions_result(Side::Overlord).is_err());
    assert_contents_equal(
        session.legal_actions(Side::Champion),
        vec![
            GameAction::PromptAction(PromptAction::MulliganDecision(MulliganDecision::Mulligan)),
            GameAction::PromptAction(PromptAction::MulliganDecision(MulliganDecision::Keep)),
        ],
    );
}

#[test]
fn keep_opening_hand() {
    let (game_id, overlord_id, champion_id) = generate_ids();
    let mut session = make_overlord_test_session(game_id, overlord_id, champion_id);
    initiate_game(&mut session);

    let response = session.click_on(overlord_id, "Keep");
    assert_eq!(0, session.user.cards.revealed_cards().len());
    assert_eq!(5, session.user.cards.hand(PlayerName::User).len());
    assert_eq!(5, session.user.cards.hand(PlayerName::Opponent).len());

    assert_eq!(0, session.opponent.cards.hand(PlayerName::User).len());
    assert_eq!(5, session.opponent.cards.hand(PlayerName::Opponent).len());
    assert_eq!(5, session.opponent.cards.revealed_cards().len());

    assert_snapshot!(Summary::summarize(&response));
}

#[test]
fn mulligan_opening_hand() {
    let (game_id, overlord_id, champion_id) = generate_ids();
    let mut session = make_overlord_test_session(game_id, overlord_id, champion_id);
    initiate_game(&mut session);

    let response = session.click_on(overlord_id, "Mulligan");
    assert_snapshot!(Summary::summarize(&response));

    assert_eq!(0, session.user.cards.revealed_cards().len());
    assert_eq!(5, session.user.cards.hand(PlayerName::User).len());
    assert_eq!(5, session.user.cards.hand(PlayerName::Opponent).len());

    assert_eq!(0, session.opponent.cards.hand(PlayerName::User).len());
    assert_eq!(5, session.opponent.cards.hand(PlayerName::Opponent).len());
    assert_eq!(5, session.opponent.cards.revealed_cards().len());
}

#[test]
fn both_keep_opening_hands() {
    let (game_id, overlord_id, champion_id) = generate_ids();
    let mut session = make_overlord_test_session(game_id, overlord_id, champion_id);
    initiate_game(&mut session);

    session.click_on(overlord_id, "Keep");
    let response = session.click_on(champion_id, "Keep");
    assert_snapshot!(Summary::summarize(&response));

    assert_eq!(5, session.user.this_player.mana());
    assert_eq!(5, session.user.other_player.mana());
    assert_eq!(5, session.opponent.this_player.mana());
    assert_eq!(5, session.opponent.other_player.mana());

    assert_eq!(3, session.user.this_player.actions());
    assert_eq!(0, session.user.other_player.actions());
    assert_eq!(0, session.opponent.this_player.actions());
    assert_eq!(3, session.opponent.other_player.actions());

    assert!(session.dusk());
}

/// Creates a [TestSession] for the Overlord player. Both players have their
/// decks populated, but neither has submitted a 'new game' request.
fn make_overlord_test_session(
    game_id: GameId,
    overlord_id: PlayerId,
    champion_id: PlayerId,
) -> TestSession {
    initialize::run();
    let overlord_deck = Deck {
        index: DeckIndex::new(0),
        name: "Overlord".to_string(),
        owner_id: overlord_id,
        side: Side::Overlord,
        identity: CardName::TestOverlordIdentity,
        cards: hashmap! {CardName::TestOverlordSpell => 45},
    };
    let champion_deck = Deck {
        index: DeckIndex::new(0),
        name: "Champion".to_string(),
        owner_id: champion_id,
        side: Side::Champion,
        identity: CardName::TestChampionIdentity,
        cards: hashmap! {CardName::TestChampionSpell => 45},
    };

    let database = FakeDatabase {
        generated_game_id: Some(game_id),
        game: None,
        players: hashmap! {
            overlord_id => PlayerData {
                id: overlord_id,
                state: None,
                decks: vec![overlord_deck.clone(), champion_deck.clone()],
                adventure: None,
                collection: hashmap! {},
                tutorial: TutorialData::default()
            },
            champion_id => PlayerData {
                id: champion_id,
                state: None,
                decks: vec![overlord_deck, champion_deck],
                adventure: None,
                collection: hashmap! {},
                tutorial: TutorialData::default()
            }
        },
    };

    TestSession::new(database, overlord_id, champion_id)
}

fn initiate_game(session: &mut TestSession) {
    session.perform(
        UserAction::NewGame(NewGameAction {
            deck_index: CHAMPION_DECK,
            opponent: session.user_id(),
            debug_options: Some(NewGameDebugOptions {
                deterministic: true,
                ..NewGameDebugOptions::default()
            }),
        })
        .as_client_action(),
        session.opponent_id(),
    );
    let _action2 = session
        .perform_action(
            UserAction::NewGame(NewGameAction {
                deck_index: OVERLORD_DECK,
                opponent: session.opponent_id(),
                debug_options: Some(NewGameDebugOptions {
                    deterministic: true,
                    ..NewGameDebugOptions::default()
                }),
            })
            .as_client_action(),
            session.user_id(),
        )
        .unwrap();

    session.connect(session.user_id()).unwrap();
    session.connect(session.opponent_id()).unwrap();
}
