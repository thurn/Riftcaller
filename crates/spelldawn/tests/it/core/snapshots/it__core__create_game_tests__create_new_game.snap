---
source: crates/spelldawn/tests/it/core/create_game_tests.rs
assertion_line: 53
expression: "Summary::run(&response)"
---

command_list: 
    RenderScreenOverlay: "<ScreenOverlay>"
